src/libmongoc/src/mongoc/mongoc-interrupt.c

```diff
@@ -0,0 +1,307 @@
+/*
+ * Copyright 2020-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mongoc/mongoc-errno-private.h"
+#include "mongoc/mongoc-interrupt-private.h"
+#include "mongoc/mongoc-log.h"
+#include "mongoc/mongoc-socket-private.h"
+#include "mongoc/mongoc-stream-socket.h"
+#include "mongoc/mongoc-trace-private.h"
+#include "common-thread-private.h"
+
+/* The interrupt stream is implemented in two ways.
+ * On POSIX, this uses the self-pipe trick.
+ * On Windows, this uses a pair of TCP sockets.
```

Comment:
The self-pipe trick is one such solution to interrupting calls to poll/select.

Windows does not enable `select` to call on both sockets and non-sockets, so on Windows this uses a pair of TCP sockets. (Which seemed less preferable, as system limits on created sockets could be hit sooner).

A cleaner solution is likely to utilize epoll/kqueue/iocp in our socket primitives functions. This was brought up in the design, and consensus was that that implementing those would require more research (e.g. to see if those are portable enough to support all of our platforms), and would be a riskier change overall.